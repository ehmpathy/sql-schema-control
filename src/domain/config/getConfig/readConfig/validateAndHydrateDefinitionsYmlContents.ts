import { DefinitionType, ChangeDefinition } from '../../../../types';

/*
  validate and hydrate an array of definitinos from a yml file:
  - some will be ChangeDefinition objects
  - the rest must be strings pointing to yml files
*/

export class InvalidDefinitionError extends Error {
  public explanation: string;
  constructor({ explanation, basis }: { explanation: string, basis: any }) {
    const message = `${explanation}: ${JSON.stringify(basis)}`;
    super(message);
    this.explanation = explanation;
  }
}
export const validateAndHydrateDefinitionsYmlContents = ({ contents }: { contents: any[] }) => {
  if (!contents) return []; // enables empty files
  return contents.map((content) => {
    if (typeof content === 'string') {
      if (content.slice(-4) !== '.yml') throw new InvalidDefinitionError({ explanation: 'string must be path to a .yml file', basis: content });
      return content;
    }
    if (typeof content === 'object') {
      if (!content.type) throw new InvalidDefinitionError({ explanation: 'definitions must have a type', basis: content });
      if (![DefinitionType.CHANGE].includes(content.type)) throw new InvalidDefinitionError({ explanation: 'unsupported definition type', basis: content });
      if (content.type === DefinitionType.CHANGE) return new ChangeDefinition(content);
    }
    throw new InvalidDefinitionError({ explanation: 'invliad content type: must be object or string', basis: content });
  });
};
