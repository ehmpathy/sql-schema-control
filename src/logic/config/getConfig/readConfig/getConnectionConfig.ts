import { ConnectionConfig } from '../../../../types';
/*
  we ask the user to define the path to a node module which exports a promiseConfig method
  this promiseConfig method returns the expected db connection config
*/
export const getConnectionConfig = async ({ modulePath }: { modulePath: string }): Promise<ConnectionConfig> => {
  // 1. import the module
  const module = require(modulePath);

  // 2. get the config from the module
  if (!module.promiseConfig) throw new Error('connection module must export a promiseConfig method');
  const config = await module.promiseConfig();

  // 3. validate that the config has the required attributes
  const validatedConfig = new ConnectionConfig(config); // the connection config runtime typechecks the inputs

  // 4. return the validated connection config
  return validatedConfig;
};
