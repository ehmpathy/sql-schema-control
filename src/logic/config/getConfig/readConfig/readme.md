NOTE: errors thrown at this layer pertain to syntax, not logical errors
  - e.g., a definition is missing an id
