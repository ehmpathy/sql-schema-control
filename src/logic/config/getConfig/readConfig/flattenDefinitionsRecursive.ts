import { readYmlFile } from './_utils/readYmlFile';
import { validateAndHydrateDefinitionsYmlContents } from './validateAndHydrateDefinitionsYmlContents';
import { ChangeDefinition, ResourceDefinition } from '../../../../types';
import { getReadFilePath } from './_utils/getReadFilePath';

/*
  recursivly parse, validate, and read schema control definitions:
  - if the definition is a string, then assume that string is path to yml file of defitions
    - read the file
    - read the definitions recursively
    - extend the results initial definitions array with the additional results
*/
type DefinitionObject = ChangeDefinition | ResourceDefinition;
type DefinitionInput = DefinitionObject | string;
export const flattenDefinitionsRecursive = async ({
  definitions,
  readRoot,
}: {
  definitions: DefinitionInput[];
  readRoot: string;
}): Promise<DefinitionObject[]> => {
  const arraysOfDefinitions = await Promise.all(
    definitions.map(
      async (definition): Promise<DefinitionObject[]> => {
        // if the element is a ChangeDefinition object, return it in an array
        if (definition.constructor === ChangeDefinition) return [definition as ChangeDefinition]; // array since, although its the only definition we're getting from the list entry, it still needs to be "flatten"-able
        if (definition.constructor === ResourceDefinition) return [definition as ResourceDefinition];

        // since we now know it is not a ChangeDefinition, it must be a string.
        const filePath = getReadFilePath({ readRoot, relativePath: definition as string });
        const ymlContents = await readYmlFile({ filePath });
        const subRoot = filePath
          .split('/')
          .slice(0, -1)
          .join('/'); // drop the file name from filepath
        const subDefinitions = await validateAndHydrateDefinitionsYmlContents({
          readRoot: subRoot,
          contents: ymlContents,
        });
        const subFlattenedDefinitions = await flattenDefinitionsRecursive({
          readRoot: subRoot,
          definitions: subDefinitions,
        });
        return subFlattenedDefinitions;
      },
    ),
  );
  const flattenedDefinitions = ([] as DefinitionObject[]).concat(...arraysOfDefinitions);
  return flattenedDefinitions;
};
