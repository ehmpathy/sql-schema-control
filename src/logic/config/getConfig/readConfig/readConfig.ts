import { ControlConfig } from '../../../../types';
import { readYmlFile } from './_utils/readYmlFile';
import { getReadFilePath } from './_utils/getReadFilePath';
import { validateAndHydrateDefinitionsYmlContents } from './validateAndHydrateDefinitionsYmlContents';
import { flattenDefinitionsRecursive } from './flattenDefinitionsRecursive';
import { getConnectionConfig } from './getConnectionConfig';

/*
  1. read the yml file
  2. get the connection (i.e., import from the specified connection file)
  3. get the definitions and flatten them
*/
export const readConfig = async ({ filePath }: { filePath: string }) => {
  const configDir = filePath
    .split('/')
    .slice(0, -1)
    .join('/'); // drops the file name

  // get the yml
  const contents = await readYmlFile({ filePath });

  // get the language
  if (!contents.language) throw new Error('language must be defined');
  const language = contents.language;

  if (!contents.dialect) throw new Error('dialect must be defined');
  const dialect = `${contents.dialect}`; // ensure that we read it as a string, as it could be a number

  // determine if schema control should be strict
  const strict = contents.strict === undefined ? true : contents.strict;
  if (typeof strict !== 'boolean') throw new Error('strict must be a boolean');

  // get the connection config
  if (!contents.connection) throw new Error('connection must be defined');
  const connectionPath = contents.connection;
  const connection = await getConnectionConfig({
    modulePath: getReadFilePath({ readRoot: configDir, relativePath: connectionPath }),
  }); // NOTE: we expect connection path to be relative to the config path

  // get the resource and change definitions
  if (!contents.definitions) throw new Error('definitions must be defined');
  const definitionContents = contents.definitions;
  const nestedDefinitions = await validateAndHydrateDefinitionsYmlContents({
    readRoot: configDir,
    contents: definitionContents,
  });
  const definitions = await flattenDefinitionsRecursive({ readRoot: configDir, definitions: nestedDefinitions });

  // return the results
  return new ControlConfig({
    language,
    dialect,
    strict,
    connection,
    definitions,
  });
};
