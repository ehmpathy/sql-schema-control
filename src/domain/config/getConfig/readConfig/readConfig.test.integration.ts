import { readConfig } from './readConfig';

describe('readConfig', () => {
  it('should be able to read the example config provisioned in _test_utils', async () => {
    const config = await readConfig({ filePath: `${__dirname}/_test_utils/control.yml` });
    expect(config.language).toEqual('mysql');
    expect(config.dialect).toEqual(5.7);
    expect(config.connection).toMatchObject({
      host: '__HOST__',
      port: 3306,
      schema: '__SCHEMA__',
      username: '__USERNAME__',
      password: '__PASSWORD__',
    });
    expect(config.definitions.length).toEqual(4);
    expect(config).toMatchSnapshot(); // to log an example of the output
  });
});
