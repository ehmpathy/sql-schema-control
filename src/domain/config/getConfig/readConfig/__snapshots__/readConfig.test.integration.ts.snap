// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readConfig should be able to read the example config provisioned in _test_utils 1`] = `
Object {
  "connection": ConnectionConfig {
    "host": "__HOST__",
    "password": "__PASSWORD__",
    "port": 3306,
    "schema": "__SCHEMA__",
    "username": "__USERNAME__",
  },
  "definitions": Array [
    ChangeDefinition {
      "id": "first_change",
      "path": "path/to/some/sql.sql",
      "reappliable": undefined,
      "type": "change",
    },
    ChangeDefinition {
      "id": "init_20190619_1",
      "path": "init/service_user.sql",
      "reappliable": false,
      "type": "change",
    },
    ChangeDefinition {
      "id": "data_20190619_1",
      "path": "data/data_sources.sql",
      "reappliable": false,
      "type": "change",
    },
    ChangeDefinition {
      "id": "procedures_20190619_1",
      "path": "procedures/awesome_sproc.sql",
      "reappliable": true,
      "type": "change",
    },
  ],
  "dialect": 5.7,
  "language": "mysql",
}
`;
